### PR TITLE
hyperstart: Setup the network inside the VM

### PR DESCRIPTION
**Pull request by sboeuf**
Expected network has to be described to the agent running inside the
VM so as to set it up properly. Hyperstart agent implementation needs
to do that during the starting of the pod because network description
is passed through the STARTPOD command.

### CODE REVIEW

**Comment by dlespiau** [comment]
Urgh! the PR I just merged was targeting a branch that was already a PR (this one). And now pullapprove fails on the merge commit...

Probably not a great idea to do that :) a way to fix it would be to rework that branch to include all commits.

**Comment by sboeuf** [comment]
@dlespiau I have re-submitted this PR so that we have only one commit (no merge commit included). Once it will be merged, I will submit the CNM fix again.

File: cni.go [hunk 1]
```diff
@@ -37,6 +37,14 @@ func (n *cni) addVirtInterfaces(networkNS *NetworkNamespace) error {
 			return err
 		}
 
+		// Interface index found by CNI plugin has to be updated by adding an offset
+		// of 1 because it starts to count interfaces indexes from 0 instead 1.
+		// Network interface standards expect interface indexes to be count from 1,
+		// zero being never used.
+		for i := range result.IPs {
+			result.IPs[i].Interface = result.IPs[i].Interface + 1
```
Comment (dlespiau) [inline cni.go]: optional: += 1 works :)

File: network.go [hunk 2]
```diff
@@ -390,6 +390,50 @@ func createNetworkEndpoints(numOfEndpoints int) (endpoints []Endpoint, err error
...
+		return nil
+	})
+	if err != nil {
+		return []net.Interface{}, err
+	}
+
+	return ifaces, err
+}
+
+func getIPNetFromEndpoint(endpoint Endpoint, netIfaces []net.Interface) (net.IPNet, error) {
```
Comment (dlespiau) [inline network.go]: Not familiar with that code at all, but I had a question: does this work when multiple IP addresses are bound to the same interface?

File: hyperstart.go [hunk 3]
```diff
@@ -122,6 +122,25 @@ type hyperstartProxyCmd struct {
 	token   string
 }
 
+// HyperPod is the structure corresponding to the format expected
+// by hyperstart to start the pod on the guest.
+type HyperPod struct {
```
Comment (dlespiau) [inline hyperstart.go]: Seems to be missing the routes.

File: hyperstart.go [hunk 4]
```diff
@@ -122,6 +122,25 @@ type hyperstartProxyCmd struct {
...
+	Hostname   string                `json:"hostname"`
+	Containers []hyperJson.Container `json:"containers,omitempty"`
+	Interfaces []HyperNetIface       `json:"interfaces,omitempty"`
+	ShareDir   string                `json:"shareDir"`
+}
+
+// HyperNetIface is the structure corresponding to the format expected
+// by hyperstart to setup the network on the guest.
+type HyperNetIface struct {
```
Comment (dlespiau) [inline hyperstart.go]: missing MTU

File: hyperstart.go [hunk 5]
```diff
@@ -122,6 +122,25 @@ type hyperstartProxyCmd struct {
...
+	Device    string `json:"device,omitempty"`
+	NewDevice string `json:"newDevice,omitempty"`
```
Comment (dlespiau) [inline hyperstart.go]: The JSON name is actually newDeviceName

File: hyperstart.go [hunk 6]
```diff
@@ -122,6 +122,25 @@ type hyperstartProxyCmd struct {
...
+	IPAddress string `json:"ipAddress"`
```
Comment (dlespiau) [inline hyperstart.go]: there's a ipAddresses field that we currently use in our runtime to specify multiple addresses (related to my earlier question)

**Comment by dlespiau** [comment]
An idea could be to look at https://github.com/01org/cc-oci-runtime/blob/master/src/proxy.c#L1192 to see what we do today.

File: cni.go (hunk 1 above)
Comment (sboeuf) [inline cni.go]: yep you're right !

File: hyperstart.go (hunk 5 above)
Comment (sboeuf) [inline hyperstart.go]: Oh yes that's right !

File: cni.go (hunk 1 above)
Comment (dlespiau) [inline cni.go]: and actually ++ as well!

File: network.go (hunk 2 above)
Comment (sboeuf) [inline network.go]: No we cannot do that but I don't think we want to. We have several endpoints in virtcontainers, describing the veth end point inside the netns. For each of these endpoints, there is only one ipAddress, and we can give hyperstart as many as he wants. We don't have the framework to support this in virtcontainers right now, and I am wondering about the use cases.

File: hyperstart.go (hunk 6 above)
Comment (sboeuf) [inline hyperstart.go]: Ditto.

File: network.go (hunk 2 above)
Comment (dlespiau) [inline network.go]: ok, that makes more sense. Not totally sure if we need support for multiple IPs on a single interface, my guess is yes for CNM, and don't know for CNI. We'll find out soon enough if we start testing the different plugins.

Comment (sboeuf) [inline network.go]: Oh no wait, we can do that, just adding each ipAddress found for the same interface index.

File: hyperstart.go (hunk 4 above)
Comment (sboeuf) [inline hyperstart.go]: done

File: hyperstart.go (hunk 3 above)
Comment (sboeuf) [inline hyperstart.go]: done

File: network.go (hunk 2 above)
Comment (sboeuf) [inline network.go]: Will be done in a follow up PR. Results held inside Endpoint.Properties has to be modified to handle this properly.

**Comment by sboeuf** [comment]
@dlespiau I addressed all your comments (including the one about several IPs for an interface ;))

**Comment by sameo** [comment]
LGTM

**Comment by sameo** [comment]
@dlespiau Does that look good to you now ?

**Pull request merged**